network/server.go

```diff
@@ -237,70 +179,291 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
+		s.dialQueue.DeleteTask(p)
+	}
+
+	// Create an instance of the discovery service
+	discoveryService := discovery.NewDiscoveryService(
+		s,
+		routingTable,
+		s.logger,
+		make(chan struct{}),
+	)
+
+	// Register a network event handler
+	if subscribeErr := s.SubscribeFn(discoveryService.HandleNetworkEvent); subscribeErr != nil {
+		return fmt.Errorf("unable to subscribe to network events, %w", subscribeErr)
+	}
+
+	// Register the actual discovery service as a valid protocol
+	s.registerDiscoveryService(discoveryService)
+
+	// Make sure the discovery service has the bootnodes in its routing table,
+	// and instantiates connections to them
+	discoveryService.ConnectToBootnodes(s.bootnodes.getBootnodes())
+
+	// Start the discovery service
+	discoveryService.Start()
+
+	// Set the discovery service reference
+	s.discovery = discoveryService
+
+	return nil
+}
+
+// registerDiscoveryService registers the discovery protocol to be available
+func (s *Server) registerDiscoveryService(discovery *discovery.DiscoveryService) {
+	grpcStream := grpc.NewGrpcStream()
+	proto.RegisterDiscoveryServer(grpcStream.GrpcServer(), discovery)
+	grpcStream.Serve()
+
+	s.RegisterProtocol(common.DiscProto, grpcStream)
+}
+
+// setupBootnodes sets up the node's bootnode connections
+func (s *Server) setupBootnodes() error {
+	// Check the bootnode config is present
+	if s.config.Chain.Bootnodes == nil {
+		return ErrNoBootnodes
+	}
+
+	// Check if at least one bootnode is specified
+	if len(s.config.Chain.Bootnodes) < MinimumBootNodes {
+		return ErrMinBootnodes
+	}
+
+	bootnodesArr := make([]*peer.AddrInfo, 0)
+	bootnodesMap := make(map[peer.ID]*peer.AddrInfo)
+
+	for _, rawAddr := range s.config.Chain.Bootnodes {
+		bootnode, err := common.StringToAddrInfo(rawAddr)
+		if err != nil {
+			return fmt.Errorf("failed to parse bootnode %s: %w", rawAddr, err)
+		}
+
+		if bootnode.ID == s.host.ID() {
+			s.logger.Info("Omitting bootnode with same ID as host", "id", bootnode.ID)
+
+			continue
+		}
+
+		bootnodesArr = append(bootnodesArr, bootnode)
+		bootnodesMap[bootnode.ID] = bootnode
+	}
+
+	// It's fine for the bootnodes field to be unprotected
+	// at this point because it is initialized once (doesn't change),
+	// and used only after this point
+	s.bootnodes = &bootnodesWrapper{
+		bootnodeArr:       bootnodesArr,
+		bootnodesMap:      bootnodesMap,
+		bootnodeConnCount: int64(len(bootnodesArr)),
+	}
+
+	return nil
+}
+
+// setupIdentity sets up the identity service for the node
+func (s *Server) setupIdentity() error {
+	// Create an instance of the identity service
+	identityService := identity.NewIdentityService(
+		s,
+		s.logger,
+		int64(s.config.Chain.Params.ChainID),
+		s.host.ID(),
+	)
+
+	// Register the identity service protocol
+	s.registerIdentityService(identityService)
+
+	// Register the network notify bundle handlers
+	s.host.Network().Notify(identityService.GetNotifyBundle())
+
+	// Set the identity service
+	s.identity = identityService
```

Comment:
If we're only setting this field and the server is not doing anything with it, then it should be removed from the struct.